### DISCUSSION TITLE
Keeloq Help Needed .... if possible...

### DISCUSSION THREAD

**Discussion by eddee54455** [arendst/Tasmota#10168] [Q&A] [answered]
Hi Guys, I am an absolute noob with Tasmota relatively speaking...

I am running quite a few Tasmota flashed Sonoffs doing weird and useful things around my plot.... One thing I would dearly like to do is use my small "gate remote" transmitters that every member of the family has, to do some outside light switching... They are unfortunately of the "Keeloq" variety and the standard RF protocols on RFrecv configuration does not seem to support it....

I am using a SRX882 and a STX882 to play around with... Both are interfaced into Sonoff Basic R2 and R3 respectively... They are working, able to send data to each other via console...

When I transmit from a "Gate Remote" transmitter, the SRX does receive the data and push it out the data line, but the Tasmota side of things does not register the data on the console... (I have it hooked up to my old scope and can see the square wave binary pattern being rec'd)

Any idea how I could get Tasmota to read this data? I am not needing the entire code word, which includes the encrypted data, simply the Serial Number and Button Status... The chipset used in the "Gate Remote" is a HCS301... I have included the data sheet for good measure, the section explaining the data layout is Figure 4-2 ....

Any assistance would be AWESOME!!

Regds
Ed

Oh, and yes, many people have said it before, but it needs to be said again: "Mr Theo Arends, you are a fantastic and generous programmer to put your hard earned time and knowledge onto the web for all of us to use - THANK YOU!!"

[HCS301 Rempte Button Fob Chipset.pdf](https://github.com/arendst/Tasmota/files/5712181/HCS301.Rempte.Button.Fob.Chipset.pdf)

**Comment by Frogmore42** [comment]
Look here https://github.com/sui77/rc-switch/issues/351
That is the upstream source of the library that Tasmota uses.  (It doesn't currently support that chipset, which is why it is not working for you.)

**Reply by eddee54455** [reply]
I am a bit of a noob in this respect... Possibly I am looking at things way too simply... The rolling code side of things is certainly outside the RF receive library... But... My simple mind says that the "static" side of the HCS301 transmission is probably not a lot different to any other RFTX fob... A start preamble, Button status, serial number.... 

What am I missing?

E

**Reply by Frogmore42** [reply]
Tasmota just uses the library.  While I did a project almost a decade ago to decode and control a simple 315MHz RF switch, the people at the link I posted have done a lot more.  There is a question there that is similar to yours.  It got no help there, but it is asking for more than what you are.  I looked there and there is not a lot of information about how the code works (but the source is there for someone who is really interested in learning what it is doing, as opposed to just getting it to do what one wants).  What you are asking is not impossible, but somebody who has the skill to do it has to want to do it.  You are more likely to find someone who can help over there.

**Comment by Jason2866** [comment, marked as answer]
Keeloq IS already supported in Tasmota. Infos can be found here https://github.com/arendst/Tasmota/pull/7788

**Comment by eddee54455** [comment]
Noted that it is supported in the send side of things, I cant seem to get
it to work on the receive sidr...

**Comment by Jason2866** [comment]
This fork of rc-switch supports keeloq https://github.com/1technophile/rc-switch

**Comment by eddee54455** [comment]
Ok, so after many hours of foraging into unfamiliar territory, I found that
the current version of RC Switch lib as supported by Julie has the Keeloq
code mostly removed and disabled... Thanks Jason for the above pointer - I
compared libs eventually and found the lower section of the RC Switch lib
as provided with Julie has a whole lot "missing" from the end of it - This
section deals with Keeloq protocol... Not being a programmer, I did my
bestest to try and do the copy/paste from technophile lib and restore the
missing bits... No go on the compile though - If there's anybody out there
who can assist with getting this section of that lib up and running again,
I would be most appreciative!

For my purposes, I do not even require the coded section of the
transmission to be decoded - I am simply looking for the button
status/serial number from the rx string - this is to control security
lighting and a few other bits and bobs around my smallholding... I have
found that the cellphone interface I wrote under NodeRed is a bit difficult
to manage in the dark or even in the bright light (old eyes and the
like...LoL)...

As to the reasons why I would like it running under Tasmota as opposed to a
separate AhDoKnow processor or the like, Geography - I have around 40 or so
Tasmotized Sonoff units dotted around about 10 000 sq m of property looking
after anything from compressors, to lighting, to hot water boiler, to grey
water recycling and controlling irrigation pumps even! It would be so
awesome to hook a rf reciever to the outlying units to do on the spot
on/off/emergency ops...

Here's Hoping!!

**Comment by Jason2866** [comment]
Use the files from this PR https://github.com/arendst/Tasmota/pull/9121 it has full keeloq support It is replaced by a other PR because the included keeloq support was removed by request of Theo (redundant since cc1101 do support too)

**Comment by eddee54455** [comment]
Hey Jason,

Thanks heaps for the reply!!

I'm a bit lost as to how and which files to use there - As I said, I am a
really dumb noob when it comes to this!!

Regds
Ed

On Sun, 27 Dec 2020 at 12:59, Jason2866 <notifications@github.com> wrote:

> Use the files from this PR #9121
> <https://github.com/arendst/Tasmota/pull/9121> it has full keeloq support
> It is replaced by a other PR because the included keeloq support was
> removed by request of Theo (redundant since cc1101 do support too)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/10168#discussioncomment-243923>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APJA6AAJK3W6ABN7QTQ7MD3SW4HRXANCNFSM4VAIDV4A>
> .
>Edit: FWIW, here are the errors when I replace the RCSwitch Library with the one from the PR9121 and run the "Build" option under PlatformIO... I am obviously way outta my depth here!!

/home/ed/Documents/Tasmota-9.2.0 Julie Custom/tasmota/xdrv_17_rcswitch.ino: In function 'void RfInit()':
/home/ed/Documents/Tasmota-9.2.0 Julie Custom/tasmota/xdrv_17_rcswitch.ino:90:53: error: 'class RCSwitch' has no member named 'getNumProtos'
       Settings.rf_protocol_mask = (1ULL << mySwitch.getNumProtos()) -1;
                                                     ^
/home/ed/Documents/Tasmota-9.2.0 Julie Custom/tasmota/xdrv_17_rcswitch.ino:92:14: error: 'class RCSwitch' has no member named 'setReceiveProtocolMask'
     mySwitch.setReceiveProtocolMask(Settings.rf_protocol_mask);
              ^
/home/ed/Documents/Tasmota-9.2.0 Julie Custom/tasmota/xdrv_17_rcswitch.ino: In function 'void CmndRfProtocol()':
/home/ed/Documents/Tasmota-9.2.0 Julie Custom/tasmota/xdrv_17_rcswitch.ino:121:55: error: 'class RCSwitch' has no member named 'getNumProtos'
         Settings.rf_protocol_mask = (1ULL << mySwitch.getNumProtos()) -1;
                                                       ^
/home/ed/Documents/Tasmota-9.2.0 Julie Custom/tasmota/xdrv_17_rcswitch.ino:132:12: error: 'class RCSwitch' has no member named 'setReceiveProtocolMask'
   mySwitch.setReceiveProtocolMask(Settings.rf_protocol_mask);
            ^
/home/ed/Documents/Tasmota-9.2.0 Julie Custom/tasmota/xdrv_17_rcswitch.ino:137:37: error: 'class RCSwitch' has no member named 'getNumProtos'
   for (uint32_t i = 0; i < mySwitch.getNumProtos(); i++) {
                                     ^
Retrieved `.pio/build/tasmota/lib124/rc-switch-dev/RCSwitch.cpp.o' from cache
*** [.pio/build/tasmota/src/tasmota.ino.cpp.o] Error 1
